daemons/execd/remoted_schemas.c

```diff
@@ -0,0 +1,229 @@
+        _exit(pcmk_rc2exitc(rc));
+    }
+
+    /* Clean up any extra schema files we downloaded from a previous cluster
+     * connection.  After the files are gone, we need to wipe them from
+     * known_schemas, but there's no opposite operation for add_schema().
+     *
+     * Instead, unload all the schemas.  This means we'll also forget about all
+     * the installed schemas as well, which means that xml_latest_schema() will
+     * fail.  So we need to load the base schemas right now.
+     */
+    clean_up_extra_schema_files();
+    crm_schema_cleanup();
+    crm_schema_init();
+
+    rc = cib->cmds->fetch_schemas(cib, &reply, xml_latest_schema(), cib_sync_call);
+    if (rc != pcmk_ok) {
+        crm_err("Could not get schema files: %s", pcmk_strerror(rc));
+        rc = pcmk_legacy2rc(rc);
+
+    } else if (reply->children != NULL) {
+        /* The returned document looks something like this:
+         * <cib_command>
+         *   <cib_calldata>
+         *     <schemas>
+         *       <schema version="pacemaker-1.1">
+         *         <file path="foo-1.1">
+         *         </file>
+         *         <file path="bar-1.1">
+         *         </file>
+         *         ...
+         *       </schema>
+         *       <schema version="pacemaker-1.2">
+         *       </schema>
+         *       ...
+         *     </schemas>
+         *   </cib_calldata>
+         * </cib_command>
+         *
+         * All the <schemas> and <schema> tags are really just there for organizing
+         * the XML a little better.  What we really care about are the <file> nodes,
+         * and specifically the path attributes and the CDATA children (not shown)
+         * of each.  We can use an xpath query to reach down and get all the <file>
+         * nodes at once.
+         *
+         * If we already have the latest schema version, or we asked for one later
+         * than what the cluster supports, we'll get back an empty <schemas> node,
+         * so all this will continue to work.  It just won't do anything.
+         */
+        crm_foreach_xpath_result(reply, "//" PCMK__XA_FILE, write_extra_schema_file, NULL);
+
+        /* Don't just crm_schema_init here because that will load the base
+         * schemas again too.  Instead just load the things we fetched.
+         */
+        pcmk__load_schemas_from_dir(PCMK__REMOTE_SCHEMA_DIR);
+    }
+
+    cib__clean_up_connection(&cib);
+    _exit(pcmk_rc2exitc(rc));
+}
+
+void
+remoted_request_cib_schema_files(void)
+{
+    int rc;
+    pid_t pid;
+
+    pid = fork();
+
+    switch (pid) {
+        case -1: {
+            rc = errno;
+            crm_warn("Could not spawn process to get schema files: %s", pcmk_rc_str(rc));
+            break;
+        }
+
+        case 0:
+            /* child */
+            get_schema_files();
+            break;
+
+        default:
+            /* parent */
+            mainloop_child_add(pid, 0, "schema-fetch", NULL, NULL);
```

Comment:
A timeout even if it's longish would help if there's some sort of recovery loop for the remote connection and the children hang for whatever reason. For completeness it might be worth tracking whether there's a child active and kill it (and clean the directory) before forking a new one.

Reply:
What kind of thing are you envisioning here?  I've not had any luck getting the client set up in any way where it can actually be killed when the timeout expires - I don't think you can kill a sleeping process (or at least, I haven't been able to), and I think if the child is waiting to hear back from the server, it'll be stuck waiting on IO which I also don't think is killable.

Reply:
If you pass a timeout here, the library will automatically handle the killing in that case. Even if the process doesn't immediately go away, it prevents it from doing anything further if it wakes up, and allows it to be reaped if the parent dies without waiting for it.

For the case where a previous child is still active when this gets called again, `mainloop_child_kill()` should work the same.

Reply:
That's not what I am seeing, unless I'm doing something wrong.  For instance, with the following patch:
```
diff --git a/daemons/execd/remoted_schemas.c b/daemons/execd/remoted_schemas.c
index 7be0d30fc..28ead991b 100644
--- a/daemons/execd/remoted_schemas.c
+++ b/daemons/execd/remoted_schemas.c
@@ -132,6 +132,8 @@ get_schema_files(void)
     cib_t *cib = NULL;
     xmlNode *reply;
 
+    sleep(10);
+
     cib = cib_new();
     if (cib == NULL) {
         _exit(ENOTCONN);
@@ -243,7 +245,7 @@ remoted_request_cib_schema_files(void)
 
         default:
             /* parent */
-            mainloop_child_add(pid, 0, "schema-fetch", NULL, get_schema_files_complete);
+            mainloop_child_add(pid, 1 * 1000, "schema-fetch", NULL, get_schema_files_complete);
             break;
     }
 }
```

I see the following in the logs on the remote node:
```
Nov 16 13:56:12.660 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
Nov 16 13:56:12.991 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
Nov 16 13:56:13.097 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
...
Nov 16 13:56:15.642 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
Nov 16 13:56:22.726 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1141)        trace: Child process 773850 (schema-fetch) exited with status 0
```

And the schema files do get downloaded.

Reply:
Yeah, kill is returning ESRCH when it is called on the group:
```
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_timeout_callback@mainloop.c:1093)         info: CHILD_TIMEOUT_CALLBACK CALLED FOR 6169 (schema-fetch)
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_kill_helper@mainloop.c:1067)      info: CHILD_KILL_HELPER CALLED FOR schema-fetch
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_kill_helper@mainloop.c:1073)      info: Kill pid 6169's group
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_kill_helper@mainloop.c:1078)      info: RETURNING -3
```

Reply:
Ahhh, the group kill only works if the child calls `setpgid(0, 0)`. Either that or adding the child with `mainloop_leave_pid_group` should work. The latter is probably what we want since we don't expect this child to fork any  children of its own.

Reply:
I think I've got this figured out, then.